go/database/mpt/proof.go

```diff
@@ -0,0 +1,116 @@
+package mpt
+
+import (
+	"github.com/Fantom-foundation/Carmen/go/common"
+)
+
+//go:generate mockgen -source proof.go -destination proof_mocks.go -package witness
+
+// RlpEncodedNode is an RLP encoded MPT node.
+type RlpEncodedNode []byte
+
+// proofDb is a database of RLP encoded MPT nodes and their hashes that represent witness proofs.
+type proofDb map[common.Hash]RlpEncodedNode
+
+// WitnessProof represents a witness proof.
+// It contains a database of MPT nodes and their hashes.
+// The proof may be verified by iterating the proof database from the
+// root hash to a terminal entry according to a Path
+// representing either an account address or a storage key.
+// The proof is valid if it can be iterated to an expected terminal entry.
+type WitnessProof proofDb
+
+// Merge merges the input witness proof into the current witness proof.
+func (p *WitnessProof) Merge(other *WitnessProof) {
+	for k, v := range *other {
+		(*p)[k] = v
+	}
+}
+
+// ExtractAccountProof extracts a sub-proof from the current witness proof.
+// It returns a copy that contains only proof for the input address and storage keys.
+// If the input address or keys are not found in the current proof, it returns an empty proof.
+func (p *WitnessProof) ExtractAccountProof(nodeSource NodeSource, root common.Hash, address common.Address, keys ...common.Key) *WitnessProof {
+	return &WitnessProof{}
+}
+
+// ExtractStorageProof extracts a sub-proof from the current witness proof.
+// It returns a copy that contains only proof for the storage key.
+// If the key is not found in the current proof, it returns an empty proof.
+func (p *WitnessProof) ExtractStorageProof(nodeSource NodeSource, storageRoot common.Hash, address common.Address, key common.Key) *WitnessProof {
```

Comment:
This one seems to be a special case of the function above for a single key.

If the intention is to produce a proof that only covers the path from the AccountNode to the Value node storing the key, the resulting proof would be kinda useless since the root hash of the storage trie is not available / accessible. You would not be able to verify the proofs for anything in such a fragment.

I would remove this function and only offer the one above.

Reply:
ok, will remove it. I was not sure if there is a use-case where only key(s) needs to be extracted